File: Data Structures & Algorithms/add-two-numbers/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* cur1 = l1;
        ListNode* cur2 = l2;
        int num, a, b = 0;
        int carry = 0;
        while (cur1 && cur2) {
            a = cur1->val;
            b = cur2->val;
            num = (a + b) % 10;
            cur2->val = num + carry;
            carry = (a + b) / 10;
            cur1 = cur1->next;
            cur2 = cur2->next;
        }
        if (cur1) {
            cur2 = l2;
            while (cur2->next) {cur2 = cur2->next;}
            cur2->next = cur1;
            cur2 = cur2->next;
            cur2->val += carry;
            if (cur2 && cur2->val >= 10) {
                while (cur2 && cur2->val >= 10) {
                carry = cur2->val / 10;
                cur2->val %= 10;
                cur2 = cur2->next;
                if (!cur2) break;
                cur2->val += carry;
                carry = 0;
                }
            }
            else {carry = 0;}
        }
        else if (cur2) {
            cur2->val += carry;
            if (cur2 && cur2->val >= 10) {
                while (cur2 && cur2->val >= 10) {
                carry = cur2->val / 10;
                cur2->val %= 10;
                cur2 = cur2->next;
                if (!cur2) break;
                cur2->val += carry;
                carry = 0;
                }
            }
            else {carry = 0;}
        }
        if (carry > 0) {
            cur2 = l2;
            while (cur2->next) {cur2 = cur2->next;}
            cur2->next = new ListNode{carry};
        }
        return l2;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        count = 0
        curr = head
        while curr:
            count += 1
            curr = curr.next
        if count == 1:
            return
        idx = count // 2
        curr = head

        for i in range(idx):
            if curr:
                if i == idx - 1:
                    temp = curr
                    curr = curr.next
                    temp.next = None
                else:
                    curr = curr.next
        
        prev = None
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp

        l = head
        while l and prev:
            temp1 = l.next
            l.next = prev
            temp2 = prev.next
            if temp1:
                prev.next = temp1
            l = temp1
            prev = temp2
            

        